models/classes/pack/ItemPack.php

```diff
@@ -128,40 +130,50 @@ public function getData()
      * Set item's assets of a given type to the pack.
      *
      * @param string $type the assets type, one of those who are supported.
-     * @param string[] $assets the list of assets' URL to load
-     *
-     * @param \tao_models_classes_service_StorageDirectory $publicDirectory
+     * @param array $assets the list of assets' URL to load
      *
      * @throw InvalidArgumentException
      */
-    public function setAssets($type, $assets, $publicDirectory)
+    public function setAssets(string $type, $assets): void
     {
-        if (!in_array($type, self::$assetTypes)) {
-            throw new InvalidArgumentException('Unknow asset type "' . $type . '", it should be either ' . implode(', ', self::$assetTypes));
-        }
         if (!is_array($assets)) {
             throw new InvalidArgumentException('Assests should be an array, "' . gettype($assets) . '" given');
         }
 
-        /**
-         * Apply active encoder immediately
-         * @var Encoding $encoder
-         */
-        $encoder = EncoderService::singleton()->get($this->assetEncoders[$type], $publicDirectory);
         foreach ($assets as $asset) {
-            if ($asset instanceof MediaAsset) {
-                $mediaSource = $asset->getMediaSource();
-                if ($mediaSource instanceof MediaSource || $mediaSource instanceof HttpSource) {
-                    $assetKey = $asset->getMediaIdentifier();
-                } else {
-                    $assetKey = $mediaSource->getBaseName($asset->getMediaIdentifier());
-                }
+            $this->setAsset($type, $asset);
+        }
+    }
+
+    /**
+     * @param string $type
+     * @param string|MediaAsset $asset
+     *
+     * @throw InvalidArgumentException
+     */
+    public function setAsset(string $type, $asset): void
+    {
+        if (!in_array($type, self::$assetTypes)) {
+            throw new InvalidArgumentException('Unknow asset type "' . $type . '", it should be either ' . implode(', ', self::$assetTypes));
+        }
+
+        /** @var Encoding $encoder */
+        $encoder = EncoderService::singleton()->get($this->assetEncoders[$type]);
+
+        if ($asset instanceof MediaAsset) {
+            $mediaSource = $asset->getMediaSource();
+
+            if ($mediaSource instanceof MediaSource || $mediaSource instanceof HttpSource) {
+                $assetKey = $asset->getMediaIdentifier();
             } else {
-                $assetKey = $asset;
+                $assetKey = $mediaSource->getBaseName($asset->getMediaIdentifier());
             }
 
-            $this->assets[$type][$assetKey] = $encoder->encode($asset);
+        } else {
+            $assetKey = $asset;
         }
```

Comment:
Can we extract this to a separate method?

Reply:
I agree. Gave a suggestion: https://github.com/oat-sa/extension-tao-item/pull/410/files#r448169600